### ISSUE TITLE
Busser and boot2docker not installing tests into the sandbox

### ISSUE THREAD

**Issue by lawrencejones** [someara/kitchen-dokken#108]
I'm having issues using kitchen-dokken with boot2docker on Mac, when using busser-serverspec as the verifier.

Dokken sees `DOCKER_HOST` as a tcp connection, so `remote_docker_host?` becomes true. From what I understand, whether a host is remote is used to determine whether it will support bind mounts to directories on the host, which in this case is an incorrect interpretation as boot2docker does, but- ignoring this for now- dokken will proceed to create a data container to serve as cache.

This all works fine, up until we attempt the verify stage. I'm going to take an snippet from `lib/kitchen/helpers.rb` so I can step through what happens at each stage.

```ruby
# Extract from lib/kitchen/helpers.rb lines 189-208
module Kitchen::Verifier::Base
  def call(state)
    create_sandbox # 1
    sandbox_dirs = Dir.glob(File.join(sandbox_path, "*"))

    instance.transport.connection(state) do |conn|
      conn.execute(install_command) # 2

      unless state[:data_container].nil? # 3
	conn.execute(init_command) # 4
	info("Transferring files to #{instance.to_str}")
	conn.upload(sandbox_dirs, config[:root_path]) # 5
	debug("Transfer complete")
      end

      conn.execute(prepare_command)
      conn.execute(run_command)
    end
  rescue Kitchen::Transport::TransportFailed => ex
    raise ActionFailed, ex.message
  end
end
```

## 1 - `create_sandbox`

This step copies all the required test files into the data container. I can watch as my tests (`suites/serverspec/sample_spec.rb`) are placed inside the data container at `/opt/verifier`, as expected.

## 2 - `conn.execute(install_command)

The install command installs busser onto the container where we'll be running our tests, which we'll call `ubuntu` from here on. This runs fine, and successfully installs busser.

## 3 - `unless state[:data_container].nil?`

We have a data container, so we enter this conditional block.

## 4 - `conn.execute(init_command)`

This is where things go very wrong. The init command for busser is `/opt/verifier/bin/busser suite cleanup`. This gets busser to remove all the suite files inside the `ubuntu` container- and it does so. Only problem is that docker has bind mounted the temporary verifier directory on my local host to `/opt/verifier` on the `ubuntu` machine, meaning this command removes the result of (1) `create_sandbox` as well as all the test files on the `ubuntu` box. There are now no test files in either the temporary sandbox directory, or on the `ubuntu` box.

## 5 - `conn.upload(sandbox_dirs, config[:root_path])`

This line now tries uploading `sandbox_dirs`, which looks something like `["/Users/lawrencejones/.dokken/verifier_sandbox/34d32f30dd-dokken-ubuntu/suites"]` to the container, via an `rsync` command. The suites folder is now empty, so we `rsync` up an empty folder. This then results in there being no test files available on the box to run as tests.

---

# Potential Solutions

So the line that causes a lot of these problems is https://github.com/someara/kitchen-dokken/blob/master/lib/kitchen/driver/dokken.rb#L205. Here we add the local sandbox directory to the binds for the `ubuntu` container, despite the fact that we assume later that the bind is not in place if we have a data container.

From what I can gather about the code, this assumption is only safe if we assume that `remote_docker_host?` being true would imply binds wouldn't take effect. This is not true with boot2docker on Mac, and as a result we end up erasing the tests we want to copy before we try to rsync the tests across.

We can fix this by guarding the bind on https://github.com/someara/kitchen-dokken/blob/master/lib/kitchen/driver/dokken.rb#L205 like so:

```ruby
ret << "#{dokken_verifier_sandbox}:/opt/verifier" if dokken_verifier_sandbox && !remote_docker_host?
```

and that does indeed fix the problem. I'm unsure if this is the right approach though, as it feels to me like the data container is an unnecessary component if we have binds available to us, and besides, for remote docker instances we shouldn't ever be attempting to bind anything.

Alternatively, we could redefine `remote_docker_host?` as:

```ruby
def remote_docker_host?
  return false if `docker info`.include?('boot2docker')
  return true if config[:docker_host_url] =~ /^tcp:/
  false
end
```

I've not PR'd either of these changes myself because I don't fully understand the reasoning behind why things are as they are, and wanted to get someone who is more familiar with dokken to verify if this reasoning makes sense. It's very possible I'm just doing something wrong here and it's a misconfiguration- in fact I hope that's the case!

**Comment by someara** [comment]
Hi.

Thanks for this detailed explanation. It helps a lot.
I'm switching to docker toolbox myself, so I'll be fixing this one soon. 
Stay tuned.

-s

**Comment by someara** [comment]
Is this fixed up for you in 2.1.10?

**Comment by lawrencejones** [comment]
Just checked and it works for me! Nice work!

I think we can consider this fixed in 2.1.10 (https://github.com/someara/kitchen-dokken/commit/e2919b7ef4b36ecd10f8315c22033ff68d9cd3e5) 🎉
